From cae3a8510282b8d062b677f03a671b304db5e1be Mon Sep 17 00:00:00 2001
From: Patrick Mooney <pmooney@pfmooney.com>
Date: Fri, 15 Feb 2019 17:18:50 +0000
Subject: [PATCH] OS-7583 bhyve goes awry on large memory systems Reviewed by:
 Robert Mustacchi <rm@joyent.com> Reviewed by: Jerry Jelinek
 <jerry.jelinek@joyent.com> Approved by: Jerry Jelinek
 <jerry.jelinek@joyent.com>

---
 .../compat/freebsd/amd64/machine/vmparam.h    | 28 +++++++++++++------
 1 file changed, 20 insertions(+), 8 deletions(-)

diff --git a/usr/src/compat/freebsd/amd64/machine/vmparam.h b/usr/src/compat/freebsd/amd64/machine/vmparam.h
index 167aa1f459..c76a3259f3 100644
--- a/usr/src/compat/freebsd/amd64/machine/vmparam.h
+++ b/usr/src/compat/freebsd/amd64/machine/vmparam.h
@@ -11,7 +11,7 @@
 
 /*
  * Copyright 2013 Pluribus Networks Inc.
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 #ifndef _COMPAT_FREEBSD_AMD64_MACHINE_VMPARAM_H_
@@ -20,14 +20,26 @@
 extern caddr_t kpm_vbase;
 extern size_t kpm_size;
 
-#define	PHYS_TO_DMAP(x)	({ 			\
-	ASSERT((uintptr_t)(x) < kpm_size);	\
-	(uintptr_t)(x) | (uintptr_t)kpm_vbase; })
+static inline uintptr_t
+phys_to_dmap(uintptr_t pa)
+{
+	ASSERT3U(pa, <, kpm_size);
+	return ((uintptr_t)kpm_vbase + pa);
+}
 
-#define	DMAP_TO_PHYS(x)	({				\
-	ASSERT((uintptr_t)(x) >= (uintptr_t)kpm_vbase);		\
-	ASSERT((uintptr_t)(x) < ((uintptr_t)kpm_vbase + kpm_size));	\
-	(uintptr_t)(x) & ~(uintptr_t)kpm_vbase; })	\
+static inline uintptr_t
+dmap_to_phys(uintptr_t kva)
+{
+	const uintptr_t base = (uintptr_t)kpm_vbase;
+
+	ASSERT3U(kva, >=, base);
+	ASSERT3U(kva, <, base + kpm_size);
+
+	return (kva - base);
+}
+
+#define	PHYS_TO_DMAP(x)	phys_to_dmap(x)
+#define	DMAP_TO_PHYS(x)	dmap_to_phys(x)
 
 
 #endif	/* _COMPAT_FREEBSD_AMD64_MACHINE_VMPARAM_H_ */
-- 
2.21.0

